### PR TITLE
SPIRVProducerPass refactor - phase2

### PR DESCRIPTION
**Pull request by sjw36**
* addSPIRVInst for creating/inserting new SPIRVInstruction
  * Creates/manages ResultID based on spv::HasResultAndType in one place (almost, nextID still used for second pass)
* Consolidated on SmallVector storage for SPIRVOperandVec (replaces SPIRVOperandList)
* Consolidated Workgroup variable creation

Enhancement for clspv issue #545

### CODE REVIEW

**Comment by sjw36** [comment]
This is a bit more extensive than phase1 but addresses my biggest concern with nextID access everywhere. ResultIDs should be captured directly from the newly created SPIRVInstruction.

Apologies for dropping this on Friday, no rush.

File: lib/SPIRVProducerPass.cpp [hunk 1]
```diff
@@ -866,15 +827,6 @@ void SPIRVProducerPass::GenerateLLVMIRInfo(Module &M, const DataLayout &DL) {
...
 
-          // Handle image type specially.
-          if (IsImageBuiltin(callee_name)) {
-            TypeMapType &OpImageTypeMap = getImageTypeMap();
-            Type *ImageTy =
-                Call->getArgOperand(0)->getType()->getPointerElementType();
-            OpImageTypeMap[ImageTy] = 0;
-            getImageTypeList().insert(ImageTy);
-          }
-
```
Comment (sjw36) [inline lib/SPIRVProducerPass.cpp:877]: Moved this to getSPIRVType under the image type.

File: lib/SPIRVProducerPass.cpp [hunk 2]
```diff
@@ -1568,18 +1521,40 @@ void SPIRVProducerPass::FindWorkgroupVars(Module &M) {
...
+
+    SPIRVID VariableID = addSPIRVInst<kGlobalVariables>(spv::OpVariable, Ops);
+
+    Ops.clear();
+    Ops << MkId(ArraySizeID) << MkNum(spv::DecorationSpecId) << MkNum(spec_id);
+    addSPIRVInst<kAnnotations>(spv::OpDecorate, Ops);
+
+    LocalArgInfo info{VariableID,  ElemTy,         ArraySizeID,
+                      ArrayTypeID, PtrArrayTypeID, spec_id};
+    LocalSpecIdInfoMap[spec_id] = info;
```
Comment (sjw36) [inline lib/SPIRVProducerPass.cpp:1557]: With SPIRV Sections the different components of Workgroup variables can be created all at once.

File: lib/SPIRVProducerPass.cpp [hunk 3]
```diff
@@ -1925,716 +1900,598 @@ spv::BuiltIn SPIRVProducerPass::GetBuiltin(StringRef Name) const {
...
+}
 
-    Ops.clear();
-    Ops << MkNum(spv::StorageClassWorkgroup) << MkId(arg_info.array_type_id);
-    SPIRVInstList.push_back(new SPIRVInstruction(
-        spv::OpTypePointer, arg_info.ptr_array_type_id, Ops));
+void SPIRVProducerPass::GenerateSPIRVTypes(LLVMContext &Context,
+                                           Module &module) {
+  for (Type *Ty : getTypeList()) {
+    getSPIRVType(Ty);
```
Comment (sjw36) [inline lib/SPIRVProducerPass.cpp:2305]: At some point types can be created on demand but many tests would need updated. So for now still gather a TypeList and generate types up front.

File: lib/SPIRVProducerPass.cpp [hunk 4]
```diff
@@ -1925,716 +1900,598 @@ spv::BuiltIn SPIRVProducerPass::GetBuiltin(StringRef Name) const {
...
+  return 0;
+}
 
+void SPIRVProducerPass::GenerateSPIRVConstants() {
+  ValueList &CstList = getConstantList();
+
+  for (uint32_t i = 0; i < CstList.size(); i++) {
+    // UniqueVector ids are 1-based.
+    Constant *Cst = cast<Constant>(CstList[i + 1]);
+    getSPIRVValue(Cst);
```
Comment (sjw36) [inline lib/SPIRVProducerPass.cpp:2490]: Same concept applies as GenerateSPIRVTypes above.

File: lib/SPIRVProducerPass.cpp [hunk 5]
```diff
@@ -112,6 +114,8 @@ enum SPIRVSection {
   kSectionCount
 };
 
+typedef uint32_t SPIRVID;
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:117]: Prefer not to introduce this typedef. Just use uint32_t.

File: lib/SPIRVProducerPass.cpp [hunk 6]
```diff
@@ -496,8 +427,40 @@ struct SPIRVProducerPass final : public ModulePass {
   // Returns true if |Arg| is called with a coherent resource.
   bool CalledWithCoherentResource(Argument &Arg);
 
+  //
+  // Primary interface for adding SPIRVInstructions to a SPIRVSection.
+  template <enum SPIRVSection TSection = kFunctions>
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:432]: Why templating over just taking an extra parameter? Seems like a waste for the compiler to generate extra functions to start with. I'd also rather be explicit everywhere than just making kFunction instructions easier to insert.

File: lib/SPIRVProducerPass.cpp [hunk 7]
```diff
@@ -496,8 +427,40 @@ struct SPIRVProducerPass final : public ModulePass {
...
+  SPIRVID addSPIRVInst(spv::Op Op, const char *V) {
+    SPIRVOperandVec Ops;
+    Ops << MkString(V);
+    return addSPIRVInst<TSection>(Op, Ops);
+  }
+
 private:
   static char ID;
+
+  Module *module;
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:462]: I dislike the mixing this introduces. It's ok to store the module pointer, but I'd rather stop passing around the module when that happens. This would ideally be changed in a separate PR.

File: lib/SPIRVProducerPass.cpp [hunk 8]
```diff
@@ -1568,18 +1521,40 @@ void SPIRVProducerPass::FindWorkgroupVars(Module &M) {
...
-    FindConstant(ConstantInt::get(IdxTy, 0));
-    FindType(IdxTy);
-    FindType(arg->getType());
+    Ops.clear();
+    Ops << MkNum(spv::StorageClassWorkgroup) << MkId(ArrayTypeID);
+    SPIRVID PtrArrayTypeID = addSPIRVInst<kTypes>(spv::OpTypePointer, Ops);
+
+    // Generate OpVariable.
+    //
+    // GIDOps[0] : Result Type ID
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:1544]: Nit: What are GIDOps?

File: lib/SPIRVProducerPass.cpp [hunk 9]
```diff
@@ -1925,716 +1900,598 @@ spv::BuiltIn SPIRVProducerPass::GetBuiltin(StringRef Name) const {
...
+  }
+  case Type::PointerTyID: {
+    PointerType *PTy = cast<PointerType>(Ty);
+    unsigned AddrSpace = PTy->getAddressSpace();
+
+    if (AddrSpace != AddressSpace::UniformConstant) {
+      auto PointeeTy = PTy->getElementType();
+      if (PointeeTy->isStructTy() &&
+          dyn_cast<StructType>(PointeeTy)->isOpaque()) {
+        // TODO(sjw): assert always an image?
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:1944]: Samplers should also be in UniformConstant.

File: lib/SPIRVProducerPass.cpp (hunk 6 above)
Comment (sjw36) [reply lib/SPIRVProducerPass.cpp:432]: Since every call uses an explicit Section the compiler can optimize for it and should be pretty insignificant for compile time since there are only 11 specializations.

File: lib/SPIRVProducerPass.cpp (hunk 7 above)
Comment (sjw36) [reply lib/SPIRVProducerPass.cpp:462]: It was a convenience in a few places but I will forgo this for now and make another PR.

File: lib/SPIRVProducerPass.cpp (hunk 8 above)
Comment (sjw36) [reply lib/SPIRVProducerPass.cpp:1544]: Copied from another location, it was wrong there also. It should just be Ops.

File: lib/SPIRVProducerPass.cpp (hunk 5 above)
Comment (sjw36) [reply lib/SPIRVProducerPass.cpp:117]: It does make more explicit what is being returned vs a random number.

File: lib/SPIRVProducerPass.cpp (hunk 7 above)
Comment (sjw36) [reply lib/SPIRVProducerPass.cpp:462]: Actually needed in getSPIRVType unless I capture the DataLayout.

Comment (alan-baker) [reply lib/SPIRVProducerPass.cpp:462]: For clarity, I was proposing a first PR to make that change, then this one.

Comment (sjw36) [reply lib/SPIRVProducerPass.cpp:462]: Working on it..

**Comment by sjw36** [comment]
Updated and rebased.

**Comment by sjw36** [comment]
I can't understand why windows is so triggered by this file. If I turn on -Werror-unused-parameter I get errors in many files in lib..

**Comment by sjw36** [comment]
Ok, this is too much. We are setting up a windows build env.

**Comment by sjw36** [comment]
Another case where turning on -Werror-shadow causes compile failures in lib directory..

**Comment by sjw36** [comment]
Apologies @alan-baker , didn't realize that was also part of the windows error.

**Comment by alan-baker** [comment]
Sorry, you'll need to rebase again @sjw36.

**Comment by sjw36** [comment]
No problem @alan-baker , should be ready now.
